### PR TITLE
Added User Terminal

### PR DESCRIPTION
**Pull request by richardphi1618**
added a skeleton for user input and executing game logic

### CODE REVIEW

File: .gitignore
Comment (A5rocks) [inline .gitignore]: Should this have a more descriptive comment, of what goes in the `test` folder?

File: library/user_term.py
Comment (A5rocks) [inline library/user_term.py]: What is a "reset machine"? Maybe explain a bit more of what this is used for in the docstring?

File: library/user_term.py
Comment (A5rocks) [inline library/user_term.py]: Should this be a `ValueError` instead? Or another error, idk.

File: library/user_term.py
Comment (A5rocks) [inline library/user_term.py:32]: I believe you can remove the outer `[` `]`.

File: main.py
Comment (A5rocks) [inline main.py:1]: These comments are extraneous, there's a predefined ordering for these and that's what isort uses.

File: main.py
Comment (A5rocks) [inline main.py]: Could this `(0, 18)` be stored as a constant at the top of the file?

File: main.py
Comment (A5rocks) [inline main.py]: Not a very descriptive comment: maybe describe what the state is used for?

File: main.py
Comment (A5rocks) [inline main.py]: What if we replace this `term.location()` with a `print(term.move_xy(*ORIGIN_OF_USER_TERM))`? That way, the things we call can use a `term.location()` context manager!

Note: I'm unsure this approach will work, but it's probably worth a shot.

File: main.py
Comment (megahirtz) [inline main.py]: Shouldn't this be saving `state.user_select_space` to `state.user_select_subgrid`? (assuming that subgrid is valid which we will implement later)

File: main.py
Comment (megahirtz) [inline main.py]: Because of the walrus operator on line 33 `val` doesn't need to be assigned here.

**Review by megahirtz — CHANGES_REQUESTED** [review]
The logic behind which subgrid is selected for the next users turn needs to be reviewed. Currently when I run main.py player 2 starts their turn in the same subgrid as the one player 1 did when it should be based on which space player 1 played in.

File: main.py
Comment (A5rocks) [inline main.py]: This can be `player_active = 1 if player_active == 2 else 2`, if you want something more concise

File: main.py
Comment (A5rocks) [inline main.py]: This is unnecessary, `term.fullscreen()` clears the screen when you exit fullscreen mode.

File: tox.ini
Comment (A5rocks) [inline tox.ini]: Could you add `W503` to this line? I forgot to :P

**Review by A5rocks — CHANGES_REQUESTED** [review]
Looks pretty good, just a couple things need changing. Hope we can get this merged in a few hours!

File: .gitignore
Comment (A5rocks) [inline .gitignore]: Could this become a more detailed comment on what this folder is used for?

File: library/game_state.py
Comment (A5rocks) [inline library/game_state.py:32]: We can use an `enum.IntEnum` for `self.next`.

```py
class NextPossibilities(enum.IntEnum):
    WAIT_FOR_READY = 10
    ...
```
then just `self.next == NextPossibilities.WAIT_FOR_READY`

File: library/game_state.py
Comment (A5rocks) [inline library/game_state.py:40]: Is this legal state? If not, should we raise an error here?

File: library/game_state.py
Comment (A5rocks) [inline library/game_state.py:105]: Is this TODO something achievable right now?

File: library/game_state.py
Comment (A5rocks) [inline library/game_state.py]: Bolded maybe?

File: library/game_state.py
Comment (A5rocks) [inline library/game_state.py]: Should this be a `not in ("X", "O")` to make it easier to change the nothing-here character?

File: library/user_term.py
Comment (A5rocks) [inline library/user_term.py]: `updated_info += [" "] * (4 - len(updated_info))` possibly?

File: main.py
Comment (A5rocks) [inline main.py]: correction: `O` (I think?)

**Review by A5rocks — CHANGES_REQUESTED** [review]
A few comments

File: library/user_term.py
Comment (richardphi1618) [reply library/user_term.py]: I got rid of it. It was more trouble than it was worth.
I figured it was unlikely the user would cause this to error and if we did see an error it would be in debugging and we would fix it.

File: main.py
Comment (richardphi1618) [reply main.py]: this didnt work as I set the init to player_active=0. I wanted to be sure to catch it on the first pass.

File: .gitignore
Comment (richardphi1618) [reply .gitignore]: this was just a junk folder I use to store snippets. feel free to remove if no good or comment as "snippet storage"

File: library/game_state.py
Comment (richardphi1618) [reply library/game_state.py:32]: I will look into this on the next go around? Will this hold up the merge?

File: library/game_state.py
Comment (richardphi1618) [reply library/game_state.py:40]: yea I agree we should catch errors here and I think that was my intention just missed the comment

File: library/game_state.py
Comment (richardphi1618) [reply library/game_state.py:105]: kinda.... I made an attempt. More needs to be done for a proper error handler.

File: library/game_state.py
Comment (richardphi1618) [reply library/game_state.py]: sure.... why not...

File: library/game_state.py
Comment (richardphi1618) [reply library/game_state.py]: sure

File: library/user_term.py
Comment (richardphi1618) [reply library/user_term.py]: oooh I do like that...

File: main.py
Comment (richardphi1618) [reply main.py]: good catch

File: library/game_state.py
Comment (A5rocks) [reply library/game_state.py:105]: I think that's good enough then lol

File: library/game_state.py
Comment (A5rocks) [reply library/game_state.py:32]: We can do it later, sure.

File: main.py
Comment (richardphi1618) [reply main.py]: I may be dense but I am not totally following your comment. I have changed things so much I dont think this is relevant anymore? Iono let me know if I need to think about this deeper.

Comment (megahirtz) [reply main.py]: Yes the line has changed from 33 to 24, but basically this `while (val := term.inkey()) != "q":` line means you don't need to declare `val = ""` here.

Comment (A5rocks) [reply main.py]: Yeah the walrus is now on line 24, and the assignment can still be removed.

**Review by megahirtz — CHANGES_REQUESTED** [review]
LGTM once my comment is resolved, let me know if my comment still needs clarification.

File: .gitignore
Comment (A5rocks) [reply .gitignore]: I'd say we should comment it as "folder for testing things about blessed".

File: main.py
Comment (richardphi1618) [reply main.py]: okay makes sense, done!

**Review by A5rocks — APPROVED** [review]

**Review by megahirtz — APPROVED** [review]

**Pull request merged**
